### ISSUE TITLE
v2 Headers middleware hijacks CORS requests when not enabled

### ISSUE THREAD

**Issue by powerman** [containous/traefik#4979]
### Do you want to request a *feature* or report a *bug*?

Bug

### Did you try using a 1.7.x configuration for the version 2.0?

- [ ] Yes
- [x] No

### What did you do?

```yml
services:
    traefik:
        deploy:
            ...
            labels:
                ...
                traefik.http.middlewares.no-cache.Headers.customResponseHeaders.Cache-Control: "no-cache, no-store, must-revalidate"
                traefik.http.middlewares.no-cache.Headers.customResponseHeaders.Pragma: "no-cache"
                traefik.http.middlewares.no-cache.Headers.customResponseHeaders.Expires: "0"
        image: "traefik:2.0.0-alpha4"
        ...
        command: >
            --api
            --global.sendAnonymousUsage=true
            --entrypoints='Name:http Address::80'
            --entrypoints='Name:https Address::443'
            --metrics
            --metrics.prometheus
            --providers
            --providers.file.directory=/etc/traefik/
            --providers.docker
            --providers.docker.swarmmode
            --providers.docker.exposedbydefault=false
            --providers.docker.watch
            --providers.docker.network=project
            --serverstransport.insecureskipverify
            --log
            --log.level=debug
            --accesslog
            --global.debug

    something:
        deploy:
            ...
            labels:
                ...
                traefik.http.routers.something-tls.middlewares: "no-cache"
        ...
```

### What did you expect to see?

Traefik shouldn't reply on CORS OPTIONS requests on it's own, it should forward such requests to backend service.

### What did you see instead?

Traefik reply on requests like
```
Origin: https://example.com
Access-Control-Request-Method: POST
Access-Control-Request-Headers: content-type
```
with this:
```
HTTP/1.1 200 OK
Access-Control-Max-Age: 0
Content-Length: 0
Date: Tue, 18 Jun 2019 20:11:56 GMT
```
If request doesn't have `Access-Control-Request-Headers:` header then request is forwarded to backend service, as it should be.

### Output of `traefik version`: (_What version of Traefik are you using?_)

```
Version:      2.0.0-alpha4
Codename:     faisselle
Go version:   go1.12.4
Built:        2019-04-17T11:38:21Z
OS/Arch:      linux/amd64
```

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

Shown above.

**Comment by dtomcej** [comment]
Hello @powerman,

Traefik itself does not implement CORS handling logic (in regards to whether to process requests/etc). It has the ability to add cors headers in the case your application does not generate them, but it still relies on the browser, or the application to implement the actual logic.

Also of note, you haven't configured traefik to add/remove CORS headers, so Traefik is oblivious to these headers, and passes them on to the backend as it receives them.

I will go ahead and close this since it is expected behavior. If you would like to discuss further on configuration, please reach out to us on slack: traefik.slack.com

Thanks!

**Comment by powerman** [comment]
You shouldn't close this, because traefik isn't oblivious to these headers, that's the issue. It does respond to such request on it's own, our backend doesn't see this request at all.

If client doesn't send `Access-Control-Request-Headers:` then traefik is deliver request to backend, as it should.

**Comment by dtomcej** [comment]
@powerman Thank you for the update. I think I now understand what you are saying:

Traefik is responding to CORS preflight requests without being configured to handle CORS requests.

To confirm:

You are configuring custom headers via the header middleware, and traefik is now responding to cors preflight requests without you enabling any cors features?

I think I found the code that is broken: https://github.com/containous/traefik/blob/v2.0/pkg/middlewares/headers/headers.go#L130

I will update my open PR (https://github.com/containous/traefik/pull/4857) to resolve this issue.

**Comment by powerman** [comment]
Yes, this is exactly what I mean.

**Comment by traefiker** [comment]
Closed by #4857.
